content/20.zksync-protocol/15.zksyncos/05.hooks-flows.md

```diff
@@ -0,0 +1,120 @@
+calldata for the `sendToL1(bytes)` method following solidity abi.
+
+Implementation of the l1 messenger system hook does 2 things: decodes the input and records the message using the system method.
+
+## Runner Flow
+
+This section describes how the bootloader interacts with the execution environments to run contract code. This section is complemented
+by [Execution Environments](/zksync-protocol/zksyncos/execution-environment).
+
+**Entrypoints**
+
+The bootloader implements (and uses) two entrypoints for code execution.
+
+The first one is `run_till_completion` from
+the [`runner`](https://github.com/matter-labs/zk_ee/blob/main/basic_bootloader/src/bootloader/runner.rs) module of the bootloader. This function
+implements the main execution loop given an initial request (either external call or deployment), which is explained in the next section.
+
+The second one,
+[`run_single_interaction`](https://github.com/matter-labs/zk_ee/blob/main/basic_bootloader/src/bootloader/run_single_interaction.rs),
+is just a simple wrapper over the previous to simplify external calls from the bootloader. It just adds the logic for starting and finishing the
+topmost execution frame and prepares the inputs for `run_till_completion`.
+
+**Runner Structure**
+
+The runner's responsibility is to coordinate calls into the execution environments. For this, the runner keeps a callstack of execution
+environment states and will be responsible of starting and finishing system frames. Frames are used to take snapshots for storage and memory to
+which the system can revert to in case of a failure.
+
+The runner is implemented as an infinite loop that dispatches the preemption reasons returned by the execution environment. As a reminder, these are:
+
+- External call request,
+- Deployment request,
+- External call completed, and
+- Deployment completed.
+
+The runner breaks out of the infinite loop after processing the completion of the initial request (when the callstack becomes empty).
+
+**Call Request**
+
+For the external call request, the bootloader needs to:
+
+1. Start frame for call.
+2. Run the call preparation in the system, this will return the callee's bytecode, transfer token value and charge gas according to the EE's policy.
+3. Create a new EE state for the call and push it to the callstack.
+4. Call into the newly created EE to start executing the frame.
+
+There's a special case in which the callee is a special address (for example, precompile or system contract). In this case the flow is similar,
+but there's no new EE. Instead, the [System Hooks](http://localhost:3000/zksync-protocol/zksyncos/hooks-flows) are used.
+
+**Deployment Request**
+
+For deployment request, the bootloader needs to:
+
+1. Start frame for deployment preparation.
+2. Call into EE to run deployment preparation. This will compute the deployed address, perform some checks and charge gas.
```

Comment:
```suggestion
2. Call into the EE to run the deployment preparation. This will compute the deployed address, perform some checks and charge gas.
```